File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& cand, int tar) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(begin(cand),end(cand));
        fun(cand, ans, temp, tar, 0);
        return ans;
    }
private:
    void fun(vector<int> &cand, vector<vector<int>> &ans, vector<int> &temp, int tar, int i) {
        if (i >= cand.size()) {
            if (tar == 0) {
                ans.push_back(temp);
            }
            return;
        }
        if (tar == 0) {
            ans.push_back(temp);
            return;
        }
        if ((tar - cand[i]) >= 0) {
            temp.push_back(cand[i]);
            fun(cand, ans, temp, tar - cand[i], i);
            temp.pop_back();
        }
        fun(cand, ans, temp, tar, i+1);
    }
};

File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, cand: List[int], target: int) -> List[List[int]]:
        res,tmp = [],[]
        def helper(i, cur):
            if i >= len(cand):
                if cur == target:
                    res.append(tmp[:])
                return
            if cur == target:
                res.append(tmp[:])
                return
            if cur+cand[i] <= target:
                tmp.append(cand[i])
                cur += cand[i]
                helper(i, cur)
                cur -= cand[i]
                tmp.pop()
            helper(i+1, cur)
        helper(0,0)
        return res
            